<img align="right" src="../../additional_data/banner_siegel.png" style="width:1100px;">

# Advanced xArray

* [**Sign up to the JupyterHub**](https://www.phenocube.org/) to run this notebook interactively from your browser
* **Compatibility:** Notebook currently compatible with the Open Data Cube environments of the University of Wuerzburg
* **Products used**: 
* **Prerequisites**:  Users of this notebook should have a basic understanding of:
    * How to run a [Jupyter notebook](01_jupyter_introduction.ipynb)
    * The basic structure of the eo2cube [satellite datasets](02_eo2cube.ipynb)
    * How to browse through the available [products and measurements](03_products_and_measurements.ipynb) of the eo2cube datacube 
    * How to [load data from the eo2cube datacube](04_loading_data_and_basic_xarray.ipynb) 

## Background

The Python library `xarray` simplifies working with labelled multi-dimension arrays. The library introduces labels in the forms of dimensions, coordinates and attributes on top of `numpy` arrays. This structure allows easier and more effective handling of remote sensing raster data in a Python environment. Therefore, it is essential to fully understand the structure of an `xarray`. A first introduction into the usage of `xarray` within the eo2cube environment was given in ["04_loading_data_and_basic_xarray"](04_loading_data_and_basic_xarray.ipynb). This notebook builds on this gained knowledge and attempts to give a deeper understanding of the `xarray` data structure of raster data. Since the `xarray.Dataset` within the datacube environment is specialised for the use of remote sensing raster data, it differs slightly from the original `xarray` library. However, if you are interested in learning more about the basic structures of the original `xarray`, have a look at this [**"introduction to xarray" notebook**](intro_to_xarray.ipynb) within the "intro_to_python" directory.
To get more information about the `xarray` package, visit the [offical documentation website](http://xarray.pydata.org/en/stable/).

## Description

This notebook introduces users to the `xarray` library within the datacube environment. It aims to deepen the understanding of the `xarray` structure as a container for remote sensing raster data. Also it introduces useful `xarray` functions to effectivly work with raster data in the eo2cube environment. Within this notebook the following topics are covered:

* Definition of the `xarray.Dataset` structure for eo2cube remote sensing data
* Access of `xarray.Dataset` dimensions, measurements and metadata
* Indexing and slicing of `xarray.Dataset`
* Application of built-in `xarray` functions for analyzing raster data

***

## Load packages

The `datacube` package is required to query the eo2cube datacube database and load the requested data. The `with_ui_cbk` function from `odc.ui` enables a progress bar when loading large amounts of data. The `xarray` and `numpy` package are needed for the different methods and analysis steps within this notebook. 

In [1]:
import datacube
from odc.ui import with_ui_cbk
import xarray as xr
import numpy as np

## Datacube connection and load data

First we connect to the datacube and load a dumy dataset from the eo2cube. For this we will use the `s2_l2a_bavaria` product. An area around Würzburg is loaded for April 2020. For more information about how to use the `dc. load()` function, check out [notebook 04](04_loading_data_and_basic_xarray.ipynb).

In [2]:
dc = datacube.Datacube(app = '05_advanced_xarray', config = '/home/datacube/.datacube.conf')

In [3]:
data = dc.load(product= "s2_l2a_bavaria",
               measurements= ["blue", "green", "red"],
               x= (9.8506165, 11.273325),
               y= (49.7352601, 50.191334),
               time= ("2020-04-01", "2020-04-07"),
               group_by = "solar_day",
               progress_cbk=with_ui_cbk())

data

<xarray.Dataset>
Dimensions:      (time: 4, x: 10310, y: 5284)
Coordinates:
  * time         (time) datetime64[ns] 2020-04-01T10:26:54 ... 2020-04-06T10:...
  * y            (y) float64 5.562e+06 5.562e+06 5.562e+06 ... 5.51e+06 5.51e+06
  * x            (x) float64 5.607e+05 5.607e+05 ... 6.638e+05 6.638e+05
    spatial_ref  int32 25832
Data variables:
    blue         (time, y, x) int16 493 302 271 355 265 ... 322 323 238 202 258
    green        (time, y, x) int16 628 394 401 498 405 ... 479 429 322 311 367
    red          (time, y, x) int16 805 459 438 600 436 ... 581 512 409 380 421
Attributes:
    crs:           EPSG:25832
    grid_mapping:  spatial_ref

## `xarray.Dataset` structure

The variable `data` is an `xarray.Dataset`. A `xarray.Dataset` is basically a dictionary structure or data container that packs the raster dataset into "dimensions", "coordinates", "data variables" and "attributes".

The **"dimensions"** represent the absolute dimensions of the data, i.e. the amount of time steps (how many different scenes are available in the dataset) and the absolute pixel number in x and y (lon and lat) direction of the dataset (how many pixels exist in x and y (lon and lat) direction).

The **"coordinates"** represent the actual values of the dimensions. These are stored in `xarray.DataArrays`. In their core, `xarray.DataArrays` are a build on raw `numpy` arrays. To see a preview of the contained values click the database symbol on the right of the `xarray.DataArray`. [The section below](#index_xarray1) focuses on how to select and actually work with the values of different `xarray.DataArrays` of a dataset.
The `time` variable within the "coordinates" section displays an array with the time steps available in the dataset. In the example of `data`, the dataset contains four time steps.
The `y` and `x` variables display the pixel coordinate values for each pixel within the spatial bounds of the dataset. In our example, the `y` array holds 5284 values and the `x` array holds  5284 values (like defined in the "dimensions"). For every unique combination of `y` and `x` coordinate a value for each measurement/band exists.
The `spatial_ref` variable contains an array with only a single value which defines the EPSG code of the CRS.

The **"data variables"** display all loaded measurements/bands of the dataset. The measurements are labeled with the band names of the raster data. The actual values of a measurement are stored in a multi-dimensional `xarray.DataArray`. This array consists of three dimension. The first dimension represents the `time` variable. In our example, the `xarray.DataArray` of a band stores four arrays in the first dimensions, each representing one time step. The second dimension represents the `y` coordinates. Therefore, a band `xarray.DataArray` in `data` stores 5248 values in the second dimensions, each representing a single pixel in the y direction. The third dimension represents the `x` coordinates. For the `data` dataset, a band `xarray.DataArray` stores 10310 values in the third dimension, each representing a single pixel in the x direction. In [the following section](#index_array2) we will learn how to select and work with the band `xarray.DataArrays` and therefore obtain a deeper understanding of the multi-dimensional `xarray.DataArray` structure.

The **"attributes"** section stores the defined CRS of the loaded data.

## Inspection of a `xarray.Dataset`

As disscused above, the `xarray.Dataset` packs the raster data into different variables ("dimensions", "coordinates", "data variables", "attributes"). These sections can be addressed individually be using the follwing syntaxes:
```python
data["measurement_name"]
```
or
```python
data.measurement_name
```

The follwing code selects the "dimensions" of a `xarray.Dataset`. The result is a dict-like container storing the dimension names with the representing absolute values.

In [40]:
data.dims

Frozen(SortedKeysDict({'time': 4, 'y': 5284, 'x': 10310}))

The following code selects the "coordinates" of a `xarray.Dataset`. Te result is a dict-like container of arrays (coordinates) taht label each point.

In [47]:
data.coords

Coordinates:
  * time         (time) datetime64[ns] 2020-04-01T10:26:54 ... 2020-04-06T10:...
  * y            (y) float64 5.562e+06 5.562e+06 5.562e+06 ... 5.51e+06 5.51e+06
  * x            (x) float64 5.607e+05 5.607e+05 ... 6.638e+05 6.638e+05
    spatial_ref  int32 25832

The following code selects the "attributes" of a `xarray.Dataset`. The result is dict holding metadata.

In [45]:
data.attrs

{'crs': 'EPSG:25832', 'grid_mapping': 'spatial_ref'}

<a id='index_xarray1'></a>
### Inspecting a `xarray.DataArray`

To inspect a `xarray.DataArray` stored in "coordinates" or "data variables" use the same syntax like above. The ability to acces every `xarray.DataArray` individually can be useful to inspecting or manipulating the data stored in the array. Like a an `xarray.Dataset`, the `xarray.DataArray` also includes the information about the data´s "dimensions", "coordinates" and "attributes". The following code selects the `xarra.DataArray`containing the data for the `red` band of the `s2_l2a_bavaria` product:

In [49]:
data.red

<xarray.DataArray 'red' (time: 4, y: 5284, x: 10310)>
array([[[ 805,  459,  438, ...,  423,  505,  796],
        [ 885,  658,  550, ...,  532,  877, 1278],
        [1074,  866,  714, ...,  919, 1372, 1410],
        ...,
        [ 928,  987, 1050, ...,  366,  496,  580],
        [1062, 1128, 1252, ...,  370,  327,  449],
        [1108, 1310, 1492, ...,  455,  397,  413]],

       [[   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        ...,
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0]],

       [[3038, 2890, 2878, ...,    0,    0,    0],
        [3216, 3038, 2932, ...,    0,    0,    0],
        [3442, 3320, 3060, ...,    0,    0,    0],
        ...,
        [ 901,  924,  995, ...,    0,    0,    0],
        [ 976, 1090, 1196, ...,    0,    0,    0],
        [1013, 1240, 1366, ...,    0,    0,    0]],

       [[ 733,  420,  469, ...,  434,  449,  569],
        [ 782,  471,  480, ...,  476,  610, 1002],
        [ 852,  732,  597, ...,  631, 1102, 1420],
        ...,
        [ 915,  963,  954, ...,  665,  764,  633],
        [ 985, 1028, 1005, ...,  418,  566,  518],
        [1017, 1178, 1362, ...,  409,  380,  421]]], dtype=int16)
Coordinates:
  * time         (time) datetime64[ns] 2020-04-01T10:26:54 ... 2020-04-06T10:...
  * y            (y) float64 5.562e+06 5.562e+06 5.562e+06 ... 5.51e+06 5.51e+06
  * x            (x) float64 5.607e+05 5.607e+05 ... 6.638e+05 6.638e+05
    spatial_ref  int32 25832
Attributes:
    units:         reflectance
    nodata:        0
    crs:           EPSG:25832
    grid_mapping:  spatial_ref

In addition to the `xarray.Dataset` the specified band `xarray.DataArray` stores some additional metadata in the "attributes" section, like `nodata` values or information about the values `units`.

As you can see, the `xarray.DataArray` is still a container for the data (band) values and its related metadata ("coordinates", "attributes"). To access the actual band values stored in the`xarray.DataArray` add the argument `.values`. This returns the raw (single- or multi-dimensional) `numpy` array holding the actual band values.

In [50]:
data.red.values

array([[[ 805,  459,  438, ...,  423,  505,  796],
        [ 885,  658,  550, ...,  532,  877, 1278],
        [1074,  866,  714, ...,  919, 1372, 1410],
        ...,
        [ 928,  987, 1050, ...,  366,  496,  580],
        [1062, 1128, 1252, ...,  370,  327,  449],
        [1108, 1310, 1492, ...,  455,  397,  413]],

       [[   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        ...,
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0],
        [   0,    0,    0, ...,    0,    0,    0]],

       [[3038, 2890, 2878, ...,    0,    0,    0],
        [3216, 3038, 2932, ...,    0,    0,    0],
        [3442, 3320, 3060, ...,    0,    0,    0],
        ...,
        [ 901,  924,  995, ...,    0,    0,    0],
        [ 976, 1090, 1196, ...,    0,    0,    0],
        [1013, 1240, 1366, ...,    0,    0,    0]],

       [[ 733,  420,  469, ...,  434,

<a id='index_array2'></a>
### Multi-dimensional `xarray.DataArray` structure

The structure of a multi-dimensional measurement/band `xarray.DataArray` consists of three dimensions. The first dimension describes the `time` and represents an array for each unique time step. To select only the values of the first time step run the code below. [The next section](#index_array3) demonstrates some other `xarray` indexing functions which are more easy and effective. However, to better understand the structure of these multi-dimensional arrays, we first make use of this method.

In [70]:
data.red[0] #data values for first time step of the red band

<xarray.DataArray 'red' (y: 5284, x: 10310)>
array([[ 805,  459,  438, ...,  423,  505,  796],
       [ 885,  658,  550, ...,  532,  877, 1278],
       [1074,  866,  714, ...,  919, 1372, 1410],
       ...,
       [ 928,  987, 1050, ...,  366,  496,  580],
       [1062, 1128, 1252, ...,  370,  327,  449],
       [1108, 1310, 1492, ...,  455,  397,  413]], dtype=int16)
Coordinates:
    time         datetime64[ns] 2020-04-01T10:26:54
  * y            (y) float64 5.562e+06 5.562e+06 5.562e+06 ... 5.51e+06 5.51e+06
  * x            (x) float64 5.607e+05 5.607e+05 ... 6.638e+05 6.638e+05
    spatial_ref  int32 25832
Attributes:
    units:         reflectance
    nodata:        0
    crs:           EPSG:25832
    grid_mapping:  spatial_ref

The `xarray.DataArray` above contains all data values of the `red` band for the first time step. Still, there are the y and x dimensions which define the values. To access the second dimension (y coordinates) run the follwing code.

In [71]:
data.red[0][0]

<xarray.DataArray 'red' (x: 10310)>
array([805, 459, 438, ..., 423, 505, 796], dtype=int16)
Coordinates:
    time         datetime64[ns] 2020-04-01T10:26:54
    y            float64 5.562e+06
  * x            (x) float64 5.607e+05 5.607e+05 ... 6.638e+05 6.638e+05
    spatial_ref  int32 25832
Attributes:
    units:         reflectance
    nodata:        0
    crs:           EPSG:25832
    grid_mapping:  spatial_ref

Notice, that within the "coordinates" section the y-coordinate only contains a single value. That is because we indexed the `red` band to the first pixel in y direction. Therefore, the resulting `xarray.DataArray` now contains all the x pixels data values in the first row (first y pixel (2nd dim)) for the first time step (1st dim).

To now select the first pixel in x-direction run the following code. This gives us a unique combination of the three dimensions `time`, `y` and `x`.

In [72]:
data.red[0][0][0]

<xarray.DataArray 'red' ()>
array(805, dtype=int16)
Coordinates:
    time         datetime64[ns] 2020-04-01T10:26:54
    y            float64 5.562e+06
    x            float64 5.607e+05
    spatial_ref  int32 25832
Attributes:
    units:         reflectance
    nodata:        0
    crs:           EPSG:25832
    grid_mapping:  spatial_ref

The resulting `xarray.DataArray` contains the unique data value for the first pixel (y[0],x[0]) of the raster dataset in the first scene (time). Based on the `xarray` structure, every data value of a measurement/band in a multidimensional vector can be assigned to a pixel by a unique combination of the three dimensions.

<a id='index_array3'></a>
## Indexing

## Recommended next steps

To continue with the beginner's guide, the following notebooks are designed to be worked through in the following order:

1. [Jupyter Notebooks](01_jupyter_introduction.ipynb)
2. [eo2cube](02_eo2cube.ipynb)
3. [Products and Measurements](03_products_and_measurements.ipynb)
4. [Loading data and introduction to xarrays](04_loading_data_and_basic_xarray.ipynb)
5. **Advanced xarrays operations (this notebook)**
6. [Plotting data](06_plotting.ipynb)
7. [Basic analysis of remote sensing data](07_basic_analysis.ipynb)
8. [Parallel processing with Dask](08_parallel_processing_with_dask.ipynb)

***

## Additional information

<font size="2">This notebook for the usage in the Open Data Cube entities of the [Department of Remote Sensing](http://remote-sensing.org/), [University of Wuerzburg](https://www.uni-wuerzburg.de/startseite/), is adapted from [Geoscience Australia](https://github.com/GeoscienceAustralia/dea-notebooks), published using the Apache License, Version 2.0. Thanks! </font>

**License:** The code in this notebook is licensed under the [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0). 
Digital Earth Australia data is licensed under the [Creative Commons by Attribution 4.0](https://creativecommons.org/licenses/by/4.0/) license.


**Contact:** If you would like to report an issue with this notebook, you can file one on [Github](https://github.com).

**Last modified:** February 2021